### DISCUSSION TITLE
Timing delay in react query loading and redux dispatch

### DISCUSSION THREAD

**Discussion by naveen1337** [TanStack/query#3413] [Q&A]
React query is amazing to maintain a loading and fetching state, but when I want to do dispatch a data to redux, I am having some loading state and redux selector timing issue, 

the loading state gets false, but it takes some slight delay timing, to select data from redux, I know why is this happening, I can fix this by maintaining a loading and isFetching state inside a redux. but is there any good ways other than that ?

**Comment by TkDodo** [comment]
Why would you need to dispatch data from react-query into redux?

**Reply by naveen1337** [reply]
I needed that, because react query data is not mutable, For performing sorting and filtering operation I need to store that inside a redux, then only I can filter and mutate the data by reselect selecors

**Reply by TkDodo** [reply]
not sure what you mean with mutable, as redux store is also immutable. There are many ways to do data transformation on top of data coming from `useQuery`. The `select` option is a very good one:

```
const { data } = useQuery(key, fn, select: (result) => result.filter(...))
```

see also: https://tkdodo.eu/blog/react-query-data-transformations

---

To be honest, there is no real use-case to copy data from react-query into redux (or anywhere else, literally). You will likely have a hard time to keep those two stores in-sync, and you won't leverage the power of react-query to its fullest if you do so. Another related read would be: https://tkdodo.eu/blog/react-query-as-a-state-manager

**Reply by naveen1337** [reply]
Thanks for your explanation. but what if I want my data : Array<Objects> to be sorted by the user input.  the react query selectors looks good, but is it  recompute the function when an state changes ?

**Reply by TkDodo** [reply]
> but is it recompute the function when an state changes ?

have you read the resources I linked above? Because it is explained there:

selectors run on every render, so if you close over some local state, this will be taken into account. You can also memoize selectors with `useCallback` to only make them run if a) data from react-query changes or b) the dependency you care about changes. So yes, selectors will work like you want them to

**Reply by mcdougald** [reply]
Hey TkDodo, 

Has anything changed since you wrote that article and this [3361](https://github.com/TanStack/query/discussions/489#discussion-3361) discussion?

I am still having a problem discussed [here](https://github.com/TanStack/query/discussions/846#discussioncomment-3481716)

Basically, I have 220 form fields and having 220 useQuery calls (with/without a select) resulted in bad performance on my page. Each field is a key in the response, if the field has a response. 

Multiple users could be updating the form at the same time, ideally my project's backend/server would be better designed for this but it's a legacy system and I'm doing what I can.

**Reply by TkDodo** [reply]
does the bad perf come from re-renders or something else? I can take a look if you show a codesandbox reproduction.
but yeah, calling `useQuery` does have a slight overhead, and so does calling `select`. Not sure if it should result in measurable per degradation with 200 subscriptions though ...

**Reply by mcdougald** [reply]
Minor follow up for anyone else. I was at about 1k observers on my page for around 10 queries. 220 was the query with the highest observer count.

Part of my problem was staying on latest react-query package v3 package that supported react 18. I saw a lot of improvements finally upgrading to @tanstack/react-query v4.

Still trying to wrap my head around how I can stitch data from 5 - 8 sources together without prop-drilling a large list or obsessive observers. I haven't tried 1k observers on v4.

I moved the queries with lots of observers to redux, wrapped some props in useDeferredValue, and got my page working decently. 

Very much an issue from legacy backend architecture & early frontend mistakes.
